plutus-playground/plutus-playground-server/app/Auth.hs

```diff
@@ -0,0 +1,315 @@
+                  :<|> "github" :> GetRedirect (Headers '[ Header "Location" Text])))
+     :<|> ("gists" :> BrowserHeader "Cookie" Text :> Get '[ JSON] [Gist])
+
+type CallbackAPI
+   = "oauth" :> "github" :> "callback" :> QueryParam "code" OAuthCode :> GetRedirect (Headers '[ Header "Set-Cookie" SetCookie, Header "Location" Text])
+
+data GithubEndpoints = GithubEndpoints
+  { _githubEndpointsAuthLocation        :: !Request
+  , _githubEndpointsAccessTokenLocation :: !Request
+  , _githubEndpointsApiBaseUrl          :: !BaseUrl
+  , _githubEndpointsCallbackUri         :: !Text
+  }
+
+-- | Config determined by Github.
+mkGithubEndpoints :: IO GithubEndpoints
+mkGithubEndpoints = do
+  _githubEndpointsAuthLocation <-
+    parseRequest "GET https://github.com/login/oauth/authorize"
+  _githubEndpointsAccessTokenLocation <-
+    parseRequest "POST https://github.com/login/oauth/access_token"
+  _githubEndpointsApiBaseUrl <- parseBaseUrl "https://api.github.com"
+  let _githubEndpointsCallbackUri = "/api/oauth/github/callback"
+  pure GithubEndpoints {..}
+
+-- | Config supplied at runtime.
+data Config = Config
+  { _configJWTSignature       :: !JWT.Signer
+  , _configRedirectUrl        :: !Text
+  , _configRedirectEndpoint   :: !Text
+  , _configGithubClientId     :: !Text
+  , _configGithubClientSecret :: !Text
+  }
+
+instance FromJSON Config where
+  parseJSON =
+    withObject "config" $ \o -> do
+      _configGithubClientId <- o .: "github-client-id"
+      _configGithubClientSecret <- o .: "github-client-secret"
+      _configJWTSignature <- JWT.hmacSecret <$> o .: "jwt-signature"
+      _configRedirectUrl <- o .: "redirect-url"
+      _configRedirectEndpoint <- o .: "redirect-endpoint"
+      pure Config {..}
+
+hSessionIdCookie :: Text
+hSessionIdCookie = "sessionId"
+
+-- | The JWT key we can lookup for the Github token's value.
+githubTokenClaim :: Text
+githubTokenClaim = "github-token"
+
+redirect ::
+     (Applicative m, ToHttpApiData loc)
+  => loc -- ^ what to put in the 'Location' header
+  -> m (Headers '[ Header "Location" loc] NoContent)
+redirect a = pure $ addHeader a NoContent
+
+githubRedirect ::
+     Applicative m
+  => GithubEndpoints
+  -> Config
+  -> m (Headers '[ Header "Location" Text] NoContent)
+githubRedirect GithubEndpoints {..} Config {..} = redirect githubRedirectUrl
+  where
+    githubRedirectUrl :: Text
+    githubRedirectUrl =
+      Text.pack .
+      show .
+      getUri .
+      setQueryString
+        [ ( "redirect_uri"
+          , Just $ encodeUtf8 $ _configRedirectUrl <> _configRedirectEndpoint)
+        , ("scope", Just oauthScopes)
+        , ("client_id", Just $ encodeUtf8 _configGithubClientId)
+        ] $
+      _githubEndpointsAuthLocation
+    oauthScopes = "gist"
+
+twoWeeks :: NominalDiffTime
+twoWeeks = 60 * 60 * 24 * 7 * 2
+
+authStatus ::
+     (MonadNow m, MonadLogger m) => Config -> Maybe Text -> m AuthStatus
```

Comment:
`cookieHeader` seems to be passed around as `Maybe Text` a few times, feels wrong

Reply:
My thinking here is there are some places where no cookie right way to call it - if you aren't logged in for example - so I wanted to push the "no cookie" handling down into a single place, in `extractGithubToken`.